heft-plugins/heft-webpack5-plugin/src/Webpack5Plugin.ts

```diff
@@ -148,131 +164,241 @@ export default class Webpack5Plugin implements IHeftTaskPlugin<IWebpack5PluginOp
     return this._webpackConfiguration;
   }
 
+  private async _loadWebpackAsync(): Promise<typeof TWebpack> {
+    if (!this._webpack) {
+      // Allow this to fail if webpack is not installed
+      this._webpack = await import(WEBPACK_PACKAGE_NAME);
+    }
+    return this._webpack!;
+  }
+
+  private async _getWebpackCompilerAsync(
+    taskSession: IHeftTaskSession,
+    webpackConfiguration: IWebpackConfiguration
+  ): Promise<ExtendedCompiler | ExtendedMultiCompiler> {
+    if (!this._webpackCompiler) {
+      const webpack: typeof TWebpack = await this._loadWebpackAsync();
+      taskSession.logger.terminal.writeLine(`Using Webpack version ${webpack.version}`);
+      this._webpackCompiler = Array.isArray(webpackConfiguration)
+        ? webpack.default(webpackConfiguration) /* (webpack.Compilation[]) => MultiCompiler */
+        : webpack.default(webpackConfiguration); /* (webpack.Compilation) => Compiler */
+    }
+    return this._webpackCompiler;
+  }
+
   private async _runWebpackAsync(
     taskSession: IHeftTaskSession,
     heftConfiguration: HeftConfiguration,
-    webpackConfiguration: IWebpackConfiguration | undefined,
-    serveMode: boolean,
-    watchMode: boolean
+    options: IWebpackPluginOptions
   ): Promise<void> {
-    const logger: IScopedLogger = taskSession.logger;
+    this._validateEnvironmentVariable(taskSession);
+    if (taskSession.parameters.watch || this._serve) {
+      // Should never happen, but just in case
+      throw new InternalError('Cannot run Webpack in compilation mode when watch mode is enabled');
+    }
 
+    // Load the config and compiler, and return if there is no config found
+    const webpackConfiguration: IWebpackConfiguration | undefined = await this._getWebpackConfigurationAsync(
+      taskSession,
+      heftConfiguration,
+      options
+    );
     if (!webpackConfiguration) {
-      logger.terminal.writeLine('Webpack configuration not provided, running Webpack skipped');
       return;
     }
+    const compiler: ExtendedCompiler | ExtendedMultiCompiler = await this._getWebpackCompilerAsync(
+      taskSession,
+      webpackConfiguration
+    );
+    taskSession.logger.terminal.writeLine('Running Webpack compilation');
+
+    // Run the webpack compiler
+    let stats: TWebpack.Stats | TWebpack.MultiStats | undefined;
+    try {
+      stats = await LegacyAdapters.convertCallbackToPromise(
+        (compiler as ExtendedCompiler).run.bind(compiler)
+      );
+      await LegacyAdapters.convertCallbackToPromise(compiler.close.bind(compiler));
```

Comment:
Do we know if we still need to call close() on error?

Reply:
Unsure. This was a part of the original implementation. Errors in webpack never _seemed_ to require this to be called. The docs say the use is to finish low-priority work before `run` can be called again. However, we don't call it a second time, and we don't really care about work after a failure, so I think we're probably fine leaving this as-is